### DISCUSSION TITLE
Welcome to AnyLogic-Pypeline Discussions!

### DISCUSSION THREAD

**Discussion by t-wolfeadam** [the-anylogic-company/AnyLogic-Pypeline#10] [General]
## 👋 Welcome!
As GitHub has "Discussions" as a new feature, everything that isn't an explicit issue should be posted here. This includes:
* ❔ &nbsp;Questions about capabilities 
* &nbsp;❕ &nbsp; Ideas about new features
* 💬 Other general comments

**Comment by selinozen** [comment]
Hello, 
I want to generate random numbers in python and send them to Anylogic source or queue block as an input. However I do not know where to start or even if I write a code which results random numbers as I want, I have no idea how to send them to Anylogic and which gap needed to fill with my .py document.

**Reply by selinozen** [reply]
Is there any handbook which explains the steps gradually? I watched you webinar and still keep going on searching but after setting PyCommunicator up, it is quite confusing to solve the workflow.

**Reply by t-wolfeadam** [reply]
Hi, the README has a [quickstart](https://github.com/t-wolfeadam/AnyLogic-Pypeline#quick-start-guide) tutorial that explains very simply how to run some python code and get results back from it. The Wiki's [Basic Functions page](https://github.com/t-wolfeadam/AnyLogic-Pypeline/wiki/Usage-(basic)) and [Recommended Workflow](https://github.com/t-wolfeadam/AnyLogic-Pypeline/wiki/Recommended-Workflow) pages may also be useful. 

If you still have Qs after following/reviewing these, feel free to create a new discussion (green button on the [discussions](https://github.com/t-wolfeadam/AnyLogic-Pypeline/discussions) page) and upload your alp file/any python code and I can help further.

**Reply by selinozen** [reply]
**Thank you for the answers!** 

I have more questions and i think I will have more:)

 1. I traced the steps that you've directed me to and I could run a .py document which generates exponentially distributed random values with a working block. However, the "get x" block (as you provided in Passing Data Types document) could not give me the value which I coded on my .py document. In Python, I defined "x" in order to be found for "get x " block and I wrote the block's Action gap:
double xValue = pyCommunicator.runResults(double.class, "x");
traceln(xValue);

2. I still do not know how to keep and to be read my generated data in order to use for "source" or "queue" block. I want to share my basic ATM example with the drive link below of my questions. I am confused as to where to place my random generated data in order to use as inputs. I tried to use the code in the gap of "Interarrival Times" and there was an error occured immediately. I also tried Actions part of the "source" block and I got an error again naturally. My intention is for the blocks to read the data I am giving.
3. Isn't it possible to call a .py document which is written without def() functions. Here is an example:
Does it have to be like: 
```
import numpy as np

def generate_exponential_random(rate):
    return np.random.exponential(scale=1/rate, size=100)

x=generate_exponential_random(0.5)
print(x)

```

instead of this:
```

import numpy as np
#generating 10 numbers with rate 1
x=np.random.exponential(1,10)
print(x)
```
https://drive.google.com/drive/folders/1ZviYCrfdONwOcTW1EjARsCkj8ynMFCFR?usp=share_link

**Reply by t-wolfeadam** [reply]
Ah, thank you for sharing all of that, it really helped to see what you're trying to do. 

1 & 2. The issue with your current code is that you're generating an **_array_** of 100 random numbers at once and the "Interarrival time" field only takes one number at a time. Your current example is simple enough that it'd be much cleaner if you just generated one number at a time. You wouldn't need the Python file at all in this case, as you could just do something like:

```java
double rate = 2; // defined somewhere
pyCommunicator.run("import numpy as np"); // only needed in "On startup" field
// storing as a local variable "x" here but you could put the value in the interarrival time field directly
double x = pyCommunicator.runResults(double.class, "np.random.exponential(%s, size=1)[0]", rate);
```

If you do need an array though, you'd need to store the values in some intermediary array. If you want to pass the array to AL, you'd need to also convert from a numpy array to a standard python one. Though, I highly suggest doing this in Python, feeding the values one at a time to AnyLogic, as you'd need to also handle what happens when you get to the end of the array.

3. The benefit of defining a function for AL-purposes is the same as in Python: to wrap some complex logic in a single call. In this case, you're only calling a function in that function, so you're right, it doesn't make sense.
